linux/driver/osp-iio.c

```diff
@@ -0,0 +1,1203 @@
+		.modified = 1,
+		/* Channel 2 is use for modifiers */
+		.channel2 = IIO_MOD_X,
+		.info_mask =
+		/*
+		 * Internal bias correction value. Applied
+		 * by the hardware or driver prior to userspace
+		 * seeing the readings. Typically part of hardware
+		 * calibration.
+		 */
+		0,
+		.scan_index = axis_x,
+		.scan_type = IIO_ST('s', 32, 32, 24),
+	},
+	{
+		.type = IIO_QUATERNION,
+		.modified = 1,
+		/* Channel 2 is use for modifiers */
+		.channel2 = IIO_MOD_Y,
+		.info_mask =
+		/*
+		 * Internal bias correction value. Applied
+		 * by the hardware or driver prior to userspace
+		 * seeing the readings. Typically part of hardware
+		 * calibration.
+		 */
+		0,
+		.scan_index = axis_y,
+		.scan_type = IIO_ST('s', 32, 32, 24),
+	},
+	{
+		.type = IIO_QUATERNION,
+		.modified = 1,
+		/* Channel 2 is use for modifiers */
+		.channel2 = IIO_MOD_Z,
+		.info_mask =
+		/*
+		 * Internal bias correction value. Applied
+		 * by the hardware or driver prior to userspace
+		 * seeing the readings. Typically part of hardware
+		 * calibration.
+		 */
+		0,
+		.scan_index = axis_z,
+		.scan_type = IIO_ST('s', 32, 32, 24),
+	},
+	{
+		.type = IIO_QUATERNION,
+		.modified = 1,
+		/* Channel 2 is use for modifiers */
+		.channel2 = IIO_MOD_R,
+		.info_mask =
+		/*
+		 * Internal bias correction value. Applied
+		 * by the hardware or driver prior to userspace
+		 * seeing the readings. Typically part of hardware
+		 * calibration.
+		 */
+		0,
+		.scan_index = axis_r,
+		.scan_type = IIO_ST('s', 32, 32, 24),
+	},
+	/*
+	 * Convenience macro for timestamps. 4 is the index in
+	 * the buffer.
+	 */
+	IIO_CHAN_SOFT_TIMESTAMP(4),
+};
+
+static struct iio_chan_spec press_channels[] = {
+	{
+		.type = IIO_PRESSURE,
+		.modified = 1,
+		/* Channel 2 is use for modifiers */
+		.channel2 = IIO_MOD_X,
+		.info_mask =
+		/*
+		 * Internal bias correction value. Applied
+		 * by the hardware or driver prior to userspace
+		 * seeing the readings. Typically part of hardware
+		 * calibration.
+		 */
+		0,
+		.scan_index = axis_x,
+		.scan_type = IIO_ST('s', 32, 32, 24),
+	},
+	IIO_CHAN_SOFT_TIMESTAMP(1),
+};
+struct osp_iio_sensor {
+	struct iio_dev *indio_dev;
+	struct iio_trigger *trigger;
+	struct mutex lock;
+	int sensor;
+	int private;
+	int state;
+	long long ts;
+	union OSP_SensorData data;
+};
+
+struct osp_iio_sensor *osp_sensors_and[NUM_ANDROID_SENSOR_TYPE];
+struct osp_iio_sensor *osp_sensors_prv[NUM_PRIVATE_SENSOR_TYPE];
+
+static int osp_sensor_read_raw(struct iio_dev *indio_dev,
+			      struct iio_chan_spec const *chan,
+			      int *val, int *val2, long mask);
+static int osp_sensor_write_raw(struct iio_dev *indio_dev,
+			       struct iio_chan_spec const *chan,
+			       int val, int val2, long mask);
+/*
+ * Device type specific information.
+ */
+static const struct iio_info osp_sensor_info = {
+	.driver_module = THIS_MODULE,
+	.read_raw = &osp_sensor_read_raw,
+	.write_raw = &osp_sensor_write_raw,
+	.attrs = &osp_iio_group,
+};
+
```

Comment:
can we add read_event_config & write_event_config to read / set event state(enabled/disabled) for an iio device if events are enabled for device.